File: week09/19/P4.c
#include <stdio.h>
#include <stdlib.h>

int solution(int* A, int N) {
    if (A == NULL || N == 0) return 0;
    int dp[N];
    for (int i=0; i<N; i++) dp[i] = 0;
    int minvalue = A[0];
    for (int j=1; j<N; j++) {
        minvalue = (minvalue > A[j] ? A[j] : minvalue);
        dp[j] = (A[j]-minvalue > dp[j-1] ? A[j]-minvalue : dp[j-1]);
    }
    return dp[N-1];
}

int main() {
    int A[] = {23171, 21011, 21123, 21366, 21013, 21367};
    printf("%d\n", solution(A, 6));

    return 0;
}

File: week09/19/P4.py
def solution(A):
    if not A:
        return 0
    dp = [0] * len(A)
    minvalue = A[0]
    for i in range(1, len(A)):
        minvalue = min(minvalue, A[i])
        dp[i] = max(A[i]-minvalue, dp[i-1])
    return dp[-1]

if __name__ == "__main__":
    A = [23171, 21011, 21123, 21366, 21013, 21367]
    
    print(solution(A))